src/cljs/s3_beam/client.cljs

```diff
@@ -17,13 +21,43 @@
                (.setParameterValue "file-name" fname)
                (.setParameterValue "mime-type" fmime))))
 
-(defn sign-file [server-url file ch]
-  (let [fmap    (file->map file)
-        edn-ize #(reader/read-string (.getResponseText (.-target %)))]
-    (xhr/send (signing-url server-url (:name fmap) (:type fmap))
-           (fn [res]
-             (put! ch {:f file :signature (edn-ize res)})
-             (close! ch)))))
+(defn sign-file
+  "Takes a `server-url`, either an input map (with keys :file and optionally :identifier and :key),
+  as `input-map-or-file` and a channel `ch`. Also takes a fn `edn-ize` which is for interpreting
+  the server response text into edn.
+  Takes a `key-fn` function that takes one argument (the filename) and generates a object-key to
+  use to create the file on S3.
+  Alternatively, if a :key key exists in the input-map, use that in preference to the key-fn as an object-key.
+  Sends the request to the server-url to be signed."
+  [server-url edn-ize key-fn input-map-or-file ch]
```

Comment:
How about having an optional `opts` parameter like `[input-map-or-file ch]` & `[input-map-or-file ch opts]`. Seems that this would simplify the code in a few places.